Coinbase Wallet (#27)
nftlabs/portal: guides/create-a-pack-with-typescript-and-nextjs/create-a-pack-with-typescript-and-nextjs.mdx

---
title: Create A Pack With TypeScript And Next.js
author: "Oliver Jumpertz"
authorAvatar: "/assets/portal/authors/oliver-jumpertz.jpg"
authorTwitter: "https://twitter.com/oliverjumpertz"
draft: false

> tags: [typescript, nextjs]

Reviewer: ```suggestion
tags: [pack, typescript, nextjs, react]
```

Reviewer: we add pack so we categorize it by module (always the module being the first tag we use), and react is another good tag to add.

Reviewer 2: Changed!

summary: "Easily setup and create a Metamask wallet and how to get Testnet ETH on Rinkeby"
image: "./thumbnail.png"
date: "2021-02-01"
updated: ""
---

## Introduction

In this guide, you will learn how to work with the Pack module and create a small app that lets users get packs of random NFTs within them. Just imagine a trading card game or baseball cards where booster packs are sold that contain a random collection of individual cards.

There are a few requirements for this guide that we list for you below:

- Node.js Version >= 16 (LTS)
- A MetaMask wallet or another web wallet installed

> - Some MATIC (You can get MATIC on the Mumbai Testnet of Polygon [here](https://faucet.polygon.technology/)

Reviewer: ```suggestion
- Some MATIC (You can get MATIC on the Mumbai Testnet of Polygon [here](/portal/guides/get-matic-on-polygon-mumbai-testnet-faucet/)
```

Linking our internal guide here on how to do so.

Reviewer 2: Done, good suggestion.

- Some basic experience with JavaScript and Node.js

And now without further ado, let's get into building this app!

## Set up

> The Pack module needs either an NFT Collection, a Bundle Drop, or a Bundle Collection to work. This is why you first need to create two modules to get this project going. The easiest NFT module to use with your new Pack module is the Bundle Collection, though. This is why we will go down this route in the tutorial; to give you the most straightforward experience and your new app up and running as quickly and frictionless as possible.

Reviewer: ```suggestion
The Pack module needs either an **NFT Collection**, a **Bundle Drop**, or a **Bundle Collection** to work. This is why you first need to create two modules to get this project going. The easiest NFT module to use with your new Pack module is the Bundle Collection, though. This is why we will go down this route in the tutorial; to give you the most straightforward experience and your new app up and running as quickly and frictionless as possible.
```

I like making modules bold

> Go to the Thirdweb dashboard and create a new project. Give it any name you like. In this tutorial, we will refer to the project as "Thirdweb Packs."

Reviewer: ```suggestion
Go to the thirdweb dashboard and create a new project. Give it any name you like. In this tutorial, we will refer to the project as "Thirdweb Packs."
```

thirdweb is always all lowercase, this needs to be changed everywhere.

Reviewer: I mean, don't worry about the screenshots ofc, but the text at least

Reviewer: fml I like you very much

```TypeScript:pages/index.tsx
import type { NextPage } from 'next';
import Head from 'next/head';
import styles from '../styles/Home.module.css';
import { Connect } from '../components/connect';

const Home: NextPage = () => {
  return (
    <div className={styles.container}>
      <Head>
        <title>Thirdweb packs</title>
        <meta name="description" content="Sell and open packs of NFTs" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <Connect />
    </div>
  );
};

export default Home;
```

That's a basic homepage already. It doesn't let a user do much besides connecting their wallet, but it's a good base for any other project you might have in mind. Before we continue, we need to have another word, though.

### Attention

Usually, you would have to do a little more than only let your users connect their wallets. Whenever your users make API calls, it is better to first verify that they really own a specific wallet and are who they state they are.

> Unfortunately, implementing the whole logic to sign in users takes a little more time than we have here. You can, however, [follow our guide on letting your users sign in with their wallet](https://thirdweb.com/portal/guides/sign-in-with-ethereum-using-thirdweb-connectwallet) to also implement this logic.

Reviewer: ```suggestion
Unfortunately, implementing the whole logic to sign in users takes a little more time than we have here. You can, however, [follow our guide on letting your users sign in with their wallet](/portal/guides/sign-in-with-ethereum-using-thirdweb-connectwallet) to also implement this logic.
```

internal links don't need to be prepended by `https://thirdweb.com`

Reviewer: *and shouldn't be.

When we don't prepend, we use Link from Next.js to navigate internally.

Reviewer 2: Oopsie, forgot to alter it as I've written externally.

For now, we will just assume that everything we do is safe. But **be warned** that this is **not safe** to put into production without a proper sign-in process. We can, however, now do some more preparation.

### Create NFTs And Fund Your Module

The Pack module is not a standalone module like many of the other modules we offer. It requires an NFT module to get its NFTs from. You can imagine a pack like a wrapper around existing NFTs. Like booster packs for trading cards only pack cards from an existing collection, and we offer three types of these modules:

1. The NFT Collection

> 2. The Bundle Drop Module

Reviewer: either we mention Module on the three or in neither of them

### Implement The API Route

Handling the on-demand creation of packs when a user wants to get a new one is best handled in the backend. As you already learned, the contract has access control and not everyone should be able to create packs whenever they feel like it. Your account's private keys however are golden. They are nothing you should share with the world because everyone with that key can also access everything within it. Gladly, Next.js has API route support that allows you to build backend logic into your Next app. And all your secrets live a secure life on the server, without ever being exposed to the public.

Go to `pages/api` and create a new file `pack.ts`. This is where you implement all logic necessary to create a new pack on demand.

Let's start by adding some imports and the dotenv setup.

```TypeScript:pages/api/pack.ts
import type { NextApiRequest, NextApiResponse } from 'next';
import { ThirdwebSDK } from "@3rdweb/sdk";
import { BigNumber, ethers } from 'ethers';
import { readFileSync } from 'fs';

> require('dotenv').config();

Reviewer: ```suggestion
import dotenv from 'dotenv';
dotenv.config();
```

I know we used require on other tutorials, but we need to fix those, let's be consistent going forward using import keyword

`assets` is the most important property of the options object. Here you define all NFTs from the `assetsContract` that can potentially land inside a pack. Add in at least the ids of the NFTs you created within your bundle module.

The API creates that pack on demand. And when that is successful, it transfers the pack over to the address the caller sent with their request. After that, the id of the pack and the contract address of the pack module are returned to the user.

Overall, it only takes a few lines of code to create and transfer the pack. This is the power of the Thirdweb SDK.

You can also add a few guard statements to the API to make it more secure. After all this, `pack.ts` should look similar to this:

```TypeScript:pages/api/pack.ts
import type { NextApiRequest, NextApiResponse } from 'next';
import { ThirdwebSDK } from "@3rdweb/sdk";
import { BigNumber, ethers } from 'ethers';
import { readFileSync } from 'fs';

> require('dotenv').config();

Reviewer: same thing here with require

  {address ? <Pack /> : <p>Please connect your Wallet first.</p>}
</main>
```

`index.tsx` should now look like this:

```TypeScript:pages/index.tsx
import type { NextPage } from 'next';
import Head from 'next/head';
import styles from '../styles/Home.module.css';
import { Connect } from '../components/connect';
import { Pack } from '../components/pack';
import { useWeb3 } from '@3rdweb/hooks';

const Home: NextPage = () => {
  const { address } = useWeb3();
  return (
    <div className={styles.container}>
      <Head>
        <title>Thirdweb packs</title>
        <meta name="description" content="Sell and open packs of NFTs" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <Connect />

      <main className={styles.main}>
        {address ? <Pack /> : <p>Please connect your Wallet first.</p>}
      </main>
    </div>
  );
};

export default Home;
```

And that's it! Start your new app and try it out.

Notice that the API calls usually take some time. You might want to signal that to the user.

## Just In Case You Couldn't Follow Along

If by any chance, you couldn't follow along well, we've stored the code to this guide in our very special

> examples repository. Take a look [Take a look there](https://github.com/nftlabs/examples/tree/main/typescript/create-a-pack-with-typescript-and-nextjs) and see whether this helps you to catch up again!

Reviewer: Can we link it at the start of the guide too, so people can find it before starting and then again at the end.

Reviewer 2: Did so. Mentioned it at the beginning!
